### PR TITLE
Add SSL/TLS support for REST interface

### PR DESCRIPTION
**Pull request by prashanthpai**

### CODE REVIEW

**Comment by prashanthpai** [comment]
**Example (CLI)**

```sh
$ glustercli --host https://localhost.localdomain:24007 --cacert /home/ppai/cert/cert.pem peer status
+--------------------------------------+-------+-----------------+
|                  ID                  | NAME  |    ADDRESSES    |
+--------------------------------------+-------+-----------------+
| c583f1c6-fda6-4a62-8db8-fdc8505671f9 | gd2-1 | 127.0.0.1:24008 |
+--------------------------------------+-------+-----------------+
```

```sh
$ glustercli --host https://localhost:24007 --insecure peer status
+--------------------------------------+-------+-----------------+
|                  ID                  | NAME  |    ADDRESSES    |
+--------------------------------------+-------+-----------------+
| c583f1c6-fda6-4a62-8db8-fdc8505671f9 | gd2-1 | 127.0.0.1:24008 |
+--------------------------------------+-------+-----------------+
```

**Example (REST)**:

```sh
$ curl -i -X GET https://localhost.localdomain:24007/v1/peers --cacert /home/ppai/cert/cert.pem 
HTTP/1.1 200 OK
Content-Type: application/json; charset=UTF-8
X-Request-Id: faa40b80-bf5a-43f3-9a79-047784b8472b
Date: Fri, 29 Sep 2017 09:40:33 GMT
Content-Length: 95

[{"id":"c583f1c6-fda6-4a62-8db8-fdc8505671f9","name":"gd2-1","addresses":["127.0.0.1:24008"]}]
```

```sh
$ curl -i -X GET https://localhost.localdomain:24007/v1/peers --insecure
HTTP/1.1 200 OK
Content-Type: application/json; charset=UTF-8
X-Request-Id: c7175c65-1f5c-4a49-86ce-53b626c059fa
Date: Fri, 29 Sep 2017 09:40:22 GMT
Content-Length: 95

[{"id":"c583f1c6-fda6-4a62-8db8-fdc8505671f9","name":"gd2-1","addresses":["127.0.0.1:24008"]}]
```

**glusterd2 config file entries:**

```sh
# cat 1.yaml | grep cert
client-cert-file: "/home/ppai/cert/cert.pem"
client-key-file: "/home/ppai/cert/key.pem"
```

File: config.go [hunk 1]
```diff
@@ -43,6 +43,10 @@ func parseFlags() {
 	flag.String("clientaddress", defaultClientAddress, "Address to bind the REST service.")
 	flag.String("peeraddress", defaultPeerAddress, "Address to bind the inter glusterd2 RPC service.")
 
+	// TODO: SSL/TLS only implemented for REST interface
+	flag.String("client-cert-file", "", "Certificate used for SSL/TLS connections from clients.")
```
Comment (kshlm) [inline config.go]: client-cert implies the cert is of the clients connecting to this server. IMO, a better name would be `tls-cert` and `tls-key`.

File: cli/cmd/root.go [hunk 2]
```diff
@@ -9,21 +9,28 @@ var RootCmd = &cobra.Command{
...
 )
 
 func init() {
 	// Global flags, applicable for all sub commands
 	RootCmd.PersistentFlags().BoolVarP(&flagXMLOutput, "xml", "", false, "XML Output")
 	RootCmd.PersistentFlags().BoolVarP(&flagJSONOutput, "json", "", false, "JSON Output")
 	RootCmd.PersistentFlags().StringVarP(&flagHostname, "host", "", "http://localhost:24007", "Host")
+
+	// SSL/TLS options
+	RootCmd.PersistentFlags().StringVarP(&flagCacert, "cacert", "", "", "Path to CA certificate")
```
Comment (kshlm) [inline cli/cmd/root.go]: What is the CA cert here? The cert used by GD2?

File: pkg/restclient/common.go [hunk 3]
```diff
@@ -66,15 +71,39 @@ func (c *Client) do(method string, url string, data interface{}, expectStatusCod
 	}
 	req.Header.Set("Accept", "application/json")
 	req.Header.Set("Content-Type", "application/json")
-	resp, err1 := http.DefaultClient.Do(req)
-	if err1 != nil {
-		return err1
+
+	tr := &http.Transport{
+		DisableCompression:    true,
+		DisableKeepAlives:     true,
```
Comment (kshlm) [inline pkg/restclient/common.go]: Are these settings being set because of cmux?

File: servers/rest/rest.go [hunk 4]
```diff
@@ -2,54 +2,107 @@
...
+	rest := &GDRest{
+		Routes: mux.NewRouter(),
+		server: &http.Server{},
+		stopCh: make(chan struct{}),
+	}
+
+	enableSSL := true
+	if l, err := tlsListener(m.Match(cmux.TLS())); err != nil {
+		log.WithError(err).Warn("Failed to create SSL/TLS listener. Falling back to HTTP")
+		enableSSL = false // fallback to not using TLS
```
Comment (kshlm) [inline servers/rest/rest.go]: Don't fallback to an not using TLS. If a user has requested TLS and you cannot start a TLS listener, fail.

**Pull request merged**